File: creditrating.py
import getopt
import sys
import pandas as pd
import math


class Node:

    def __init__(self, attr, split_value, label, left_node, right_node):
        self.attr = attr
        self.split_value = split_value
        self.label = label
        self.left_node = left_node
        self.right_node = right_node


def decision_tree_learning(data_list, min_leaf):
    same_data = True
    for i in range(0, len(data_list) - 1):
        list1 = data_list[i]
        list2 = data_list[i + 1]
        if (list1[0] != list2[0] or list1[1] != list2[1] or list1[2] != list2[2] or list1[3] != list2[3]
            or list1[4] != list2[4]) or (list1[5] != list2[5]):
            same_data = False
            break

    if same_data or len(data_list) <= min_leaf:
        result = get_numbers(data_list)
        highest_frequency = result[result.values == result[0]].index
        label = highest_frequency[0]

        if len(highest_frequency) > 1:
            label = 'unknown'

        return Node(None, None, label, None, None)

    best_info = choose_best(data_list)

    left_list = []
    right_list = []
    for j in range(len(data_list)):
        if float(data_list[j][best_info[0]]) <= best_info[1]:
            left_list.append(data_list[j])
        else:
            right_list.append(data_list[j])

    left_node = decision_tree_learning(left_list, min_leaf)
    right_node = decision_tree_learning(right_list, min_leaf)

    return Node(best_info[0], best_info[1], None, left_node, right_node)


def get_numbers(data_list):
    record = []
    for i, val in enumerate(data_list):
        rating = data_list[i][5]
        record.append(rating)

    result = pd.value_counts(record)
    return result


def choose_best(data_list):
    best_gain = -1
    root_info_content = compute_info_content(data_list)
    best_info = []
    for i in range(5):
        sorted_data_list = sorted(filter(lambda x: float(x[i]) < 0, data_list), key=(lambda x: x[i]), reverse=True) \
                           + sorted(filter(lambda x: float(x[i]) > 0, data_list), key=(lambda x: x[i]))

        for j in range(0, len(sorted_data_list) - 1):
            split_value = 0.5 * (float(sorted_data_list[j][i]) + float(sorted_data_list[j + 1][i]))
            left_list = []
            right_list = []

            for k in range(len(sorted_data_list)):
                if float(sorted_data_list[k][i]) <= split_value:
                    left_list.append(sorted_data_list[k])
                else:
                    right_list.append(sorted_data_list[k])

            left_info_content = compute_info_content(left_list)
            right_info_content = compute_info_content(right_list)

            remainder = left_info_content * (len(left_list) / len(sorted_data_list)) \
                        + right_info_content * (len(right_list) / len(sorted_data_list))
            gain = root_info_content - remainder

            if gain > best_gain:
                best_gain = gain
                best_info = [i, split_value]

    return best_info


def compute_info_content(data_list):
    result = get_numbers(data_list)
    list_len = len(data_list)
    res = 0.0
    for i, val in enumerate(result):
        log_value = 0.0
        probability = result[i] / list_len
        if probability != 0:
            log_value = probability * (math.log(probability) / math.log(2))
        res = res + log_value

    res = -res
    if res == -0:
        res = 0
    return res


def predict(root, data_list):
    for i in range(len(data_list)):
        node = root

        while node.label is None:
            if float(data_list[i][node.attr]) <= node.split_value:
                node = node.left_node
            else:
                node = node.right_node

        print(node.label)


if __name__ == '__main__':
    options, args = getopt.getopt(sys.argv, 'hn:w:', longopts=[])
    a = args[0]
    trainSetName = args[1]
    testSetName = args[2]
    min_leaf = int(args[3])

    f = open('./' + trainSetName)
    data = f.readlines()
    train_data_list = list()

    for i, val in enumerate(data):
        if i > 0:
            train_data_list.append(val.split())

    f = open('./' + testSetName)
    data = f.readlines()
    test_data_list = list()

    for i, val in enumerate(data):
        if i > 0:
            test_data_list.append(val.split())

    get_numbers(train_data_list)
    dtl = decision_tree_learning(train_data_list, min_leaf)
    predict(dtl, test_data_list)

    f.close()

File: creditrating.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.*;

public class creditrating {

    public static void main(String[] args) {

        String trainSetName;

        String testSetName;

        int minleaf;

        BufferedReader br = null;


        try {

            trainSetName = args[0];
            testSetName = args[1];
            minleaf = Integer.parseInt(args[2]);

            File file = new File(trainSetName);
            br = new BufferedReader(new FileReader(file));

            String textLine;

            List<String> train_text_list = new ArrayList<>();
            br.readLine();
            while ((textLine = br.readLine())!=null) {

                train_text_list.add(textLine);
            }

            List<Double[]> train_data_list = new ArrayList<>();
            for (int i = 0; i < train_text_list.size(); i++) {
                String[] ts = train_text_list.get(i).trim().split(" +");
                Double[] td = new Double[6];
                for (int j = 0; j < ts.length -1; j++) {
                    td[j] = Double.parseDouble(ts[j]);
                }

                td[5] = Double.parseDouble(ts[5].replaceAll("AAA","0").replaceAll("AA","1")
                        .replaceAll("A","2").replaceAll("BBB","3")
                        .replaceAll("BB","4").replaceAll("B","5")
                        .replaceAll("CCC","6"));

                train_data_list.add(td);
            }


//
//            for (int i = 0; i < train_data_list.get(5).length; i++) {
//                System.out.println("s[j]:"+train_data_list.get(5)[i]);
//            }
//            System.out.println("train_text_list rating:"+train_text_list.get(99).substring(train_text_list.get(99).lastIndexOf(" ")+1));
//            System.out.println("train_text_list rating:"+train_text_list.get(98).substring(train_text_list.get(98).lastIndexOf(" ")+1));
//            System.out.println("train_text_list rating:"+train_text_list.get(95).substring(train_text_list.get(95).lastIndexOf(" ")+1));

//            getNumbers(train_data_list);
//            System.out.println(getUniqueMode(train_data_list));

//            double infoContent = computeInfoContent(train_data_list);
//

//            System.out.println(train_data_list.get(train_data_list.size()-1)[0]+","+
//                    train_data_list.get(train_data_list.size()-1)[1]+","+
//                    train_data_list.get(train_data_list.size()-1)[2]+","+
//                    train_data_list.get(train_data_list.size()-1)[3]+","+
//                    train_data_list.get(train_data_list.size()-1)[4]+","+
//                    train_data_list.get(train_data_list.size()-1)[5]);




            file = new File(testSetName);
            br = new BufferedReader(new FileReader(file));
            List<String> test_text_list = new ArrayList<>();
            br.readLine();

            while ((textLine = br.readLine())!=null) {
                test_text_list.add(textLine);
            }
            List<Double[]> test_data_list = new ArrayList<>();

            for (int i = 0; i < test_text_list.size(); i++) {
                String[] ts = test_text_list.get(i).trim().split(" +");
                Double[] td = new Double[5];
                for (int j = 0; j < ts.length; j++) {
                    td[j] = Double.parseDouble(ts[j]);
                }
                test_data_list.add(td);
            }
            br.close();

            Node dtl = DTL(train_data_list, train_text_list, minleaf);

            predict(dtl, test_data_list);

        } catch (Exception e){
            e.printStackTrace();
        }
    }



    public static Node DTL(List<Double[]> dataList, List<String> textList, int minLeaf) {

        Node node = new Node();
        // if(N≤minleaf)or(yi=yj foralli,j)or(xi=xj foralli,j)then
        boolean sameData = true;
        for (int i = 0; i < textList.size() - 1; i++) {
            String text1 = textList.get(i).trim().replaceAll(" +",",");
            String text2 = textList.get(i+1).trim().replaceAll(" +",",");

            if (!(text1.substring(0,text1.length()-2)).equals(text2.substring(0,text2.length()-2)) ||
                    !(text1.substring(text1.length()-1)).equals(text2.substring(text2.length()-1))) {
                sameData = false;
                break;
            }
        }


        if ( sameData || dataList.size() <= minLeaf ) {

            node.label = getUniqueMode(dataList);
            return node;
        }

        // [attr,splitval] ← choose-split(data)
        String bestInfo = chooseBest(dataList);
        String[] info = bestInfo.split(",");
        String attr = info[0];
        Integer attrInt = Integer.parseInt(attr);
        Double splitValue = Double.parseDouble(info[1]);


        node.attr = attrInt;
        node.splitValue = splitValue;

        List<Double[]> leftList = new ArrayList<>();
        List<Double[]> rightList = new ArrayList<>();

        List<String> leftTextList = new ArrayList<>();
        List<String> rightTextList = new ArrayList<>();

        for (int i = 0; i < dataList.size(); i++) {

            if (dataList.get(i)[attrInt] <= splitValue) {
                leftList.add(dataList.get(i));
                leftTextList.add(textList.get(i));
            } else {
                rightList.add(dataList.get(i));
                rightTextList.add(textList.get(i));
            }
        }


        node.leftNode = DTL(leftList, leftTextList, minLeaf);
        node.rightNode = DTL(rightList, rightTextList, minLeaf);

        return node;
    }



    public static String getUniqueMode(List<Double[]> dataList) {

        String rating;
        int[] record = getNumbers(dataList);
        int max = -1;
        int count = 0;
        String pos = "";
        for (int i = 0; i < record.length; i++) {
            if (record[i] > max ) {
                max = record[i];
                count = 1;
                pos = i+"";
            } else if(record[i] == max){
                count++;
            }

        }
        if (count == 1)  {
            pos = pos.replaceAll("0","AAA").replaceAll("1","AA")
                        .replaceAll("2","A").replaceAll("3","BBB")
                        .replaceAll("4","BB").replaceAll("5","B")
                        .replaceAll("6","CCC");

            return pos;
        }
        else  {
            return "unknown";
        }
    }


    public static int[] getNumbers(List<Double[]> dataList) {
        int rating;
        int[] record = new int[7];
        for (int i = 0; i < dataList.size(); i++) {

            rating = dataList.get(i)[5].intValue();

            switch(rating)
            {
                case 0:
                    record[0] = record[0]+1;
                    break;
                case 1:
                    record[1] = record[1]+1;
                    break;
                case 2:
                    record[2] = record[2]+1;
                    break;
                case 3:
                    record[3] = record[3]+1;
                    break;
                case 4:
                    record[4] = record[4]+1;
                    break;
                case 5:
                    record[5] = record[5]+1;
                    break;
                case 6:
                    record[6] = record[6]+1;
                    break;
                default:
                    System.out.println("no match");
            }
        }


        return record;
    }

    public static double computeInfoContent(List<Double[]> dataList) {
        int[] record = getNumbers(dataList);
        double res = 0;
        double naN = Double.NaN;
        for (int i = 0; i < record.length; i++) {
            double logValue = 0.000000000000;
            double probability = Double.parseDouble(record[i]+"")/dataList.size();

            if (probability != 0) {
                logValue = probability * (Math.log(probability)/Math.log(2));
//                if (logValue+"" == NaN) logValue = 0;
            }
//            System.out.println("record[i]:"+record[i]+", probability:"+probability+", logValue:"+logValue);
            res = res + logValue;
        }
        res = -res;
        if (res == -0) res = 0;
//        System.out.println("res:"+res);

        return res;
    }

    public static String chooseBest(List<Double[]> dataList) {

        String bestInfo = "";
        double bestGain = -1;


        double rootInfoContent = computeInfoContent(dataList);


        for (int attr = 0; attr < 5; attr++) {

            // Sort the array x1[attr], x2[attr], . . . , xN [attr]
            int attribute = attr;
            Comparator<Double[]> comparator = ((o1, o2) -> o1[attribute].compareTo(o2[attribute]));


            Collections.sort(dataList, comparator);

            for (int j = 0; j < dataList.size() - 1; j++) {

                // splitval ← 0.5(xi[attr] + xi+1[attr])
                double splitValue = 0.5 * (dataList.get(j)[attr] + dataList.get(j+1)[attr]);

                List<Double[]> leftList = new ArrayList<>();
                List<Double[]> rightList = new ArrayList<>();

                for (int k = 0; k < dataList.size(); k++) {

                    if (dataList.get(k)[attr] <= splitValue) {
                        leftList.add(dataList.get(k));
//
                    } else {
                        rightList.add(dataList.get(k));
                    }
                }

                double leftInfoContent = computeInfoContent(leftList);
                double rightInfoContent = computeInfoContent(rightList);


                double remainder = leftInfoContent * (Double.parseDouble(leftList.size()+"")/Double.parseDouble(dataList.size()+""))
                        + rightInfoContent * (Double.parseDouble(rightList.size()+"")/Double.parseDouble(dataList.size()+""));
                double gain = rootInfoContent - remainder;

                // bestattr ← attr and bestsplitval ← splitval
                if (gain > bestGain) {
                    bestGain = gain;
                    bestInfo = attr+","+splitValue;
                }
            }
        }


        return bestInfo;
    }




    public static void predict(Node root, List<Double[]> dataList) {


        for (int i = 0; i < dataList.size(); i++) {
            Double[] data = dataList.get(i);
            Node rootNode = root;

            while (rootNode.label == null) {
                if (data[rootNode.attr] <= rootNode.splitValue) {
                    rootNode = rootNode.leftNode;
                }
                else {
                    rootNode = rootNode.rightNode;
                }

            }
            System.out.println(" " + rootNode.label);
        }
    }


}

